### ISSUE TITLE
show graphical destination signposts for highway exits

### ISSUE THREAD

**Issue by kparal** [osmandapp/Osmand#3218]
I'd like to request osmand to show [destination](https://wiki.openstreetmap.org/wiki/Key:destination) signs for highways (motorway, trunk). 
![287px-zeichen_332-20_-_ausfahrt_von_der_autobahn _stvo_1992 svg](https://cloud.githubusercontent.com/assets/755451/19872481/3f5c5518-9fba-11e6-851b-7635fd9ba1e3.png)

Currently it is sometimes very hard to pick the correct lane and exit when using osmand. Some other applications show destination signposts that look exactly as the ones you see in real life on the highway; it is then much easier to pick the correct lane and not miss your exit (I'm aware that osmand support `turn:lanes` tags, but even with that I find some complicated exits very challenging without seeing the actual signpost contents).

For example, this is from Mapfactor Navigator:
![zue0iwn](https://cloud.githubusercontent.com/assets/755451/19871895/56197b62-9fb7-11e6-985b-25e23a5f66e2.jpg)

The blue sign shows the same information as the real signpost and helps user navigation tremendously. Osmand doesn't currently display that. If you could implement that, that would be awesome.

> What version of OsmAnd are you using, on what device and Android/iOS version?

OsmAnd~ 2.4.7 from F-Droid (planning to switch to the paid version to support the project), Google Nexus 7 (Android 6) and Motorola Moto G (Android 5)

> Do you use the offline maps offered within the OsmAnd app for download, or online (tile / raster) maps?

Offline maps, but with mobile data connection.

> If you have an issue related to offline maps, tell us the exact name of the map file where the issue occurs, and its edition date.

I haven't seen destination sign anywhere with osmand, and I know mapfactor navigator displays them in my area. I also tested this by simulating user position (using osmand debugging mode) and trying this route: 
https://www.openstreetmap.org/directions?engine=osrm_car&route=49.16061%2C16.57347%3B49.15120%2C16.60092#map=15/49.1559/16.5990
That particular [highway exit](https://www.openstreetmap.org/way/5375480#map=19/49.16038/16.59476) has correct `destination` set:
```
destination=Wien;Znojmo;Mikulov
```
The [highway directly before the exit](https://www.openstreetmap.org/way/289221299#map=19/49.16062/16.59294) even has `destination:lanes` set:
```
destination:lanes=Olomouc;Uh. Hradiště;Bratislava|Olomouc;Uh. Hradiště;Bratislava|Brno-centrum;Wien;Znojmo;Mikulov
```
However, when I simulate driving through that exit, I see no such information in the osmand. There's no signpost displayed on the screen (I've found reports that osmand might be actually _saying_ some of this information, but I'm not really interested in speech, I'd like to see it displayed).

Thank you for developing a great open-source navigation application.

**Comment by polarbearing** [comment]
While osmand does not mimick an actual roadsign, it _does_ show in writing, and uses in TTS, the ```destination``` and ```destination:ref``` information. So far it does not yet use ```destination:lanes```, IIRC. Lane recommendation is given according to the turn, however. 

What advantage does a big blue signpost-like visualisation bring over the concise form that is used currently? It just covers big parts of the map that might be useful for planning the turn.

**Comment by kparal** [comment]
That's interesting, I don't see that. I don't remember seeing ` >> destination` part in the top field, ever.

The example I picked in the original description was a bit complex, so I selected a better one. Here it is:
https://www.openstreetmap.org/directions?engine=osrm_car&route=49.16420%2C16.55566%3B49.17005%2C16.55316#map=16/49.1675/16.5539
The [motorway link](https://www.openstreetmap.org/way/5375483#map=18/49.16767/16.54838) in question has these properties:
```
destination 	Svitavy;Výstaviště
destination:int_ref 	E 461
destination:ref 	23
highway 	motorway_link
int_ref 	E 461
```

However, when I do the drive simulation, I don't see anything except `destination:ref`. I definitely don't see `destination` anywhere. (And `destination:int_ref` also doesn't seem to be supported).
![screenshot_20161101-122331](https://cloud.githubusercontent.com/assets/755451/19890389/b651c56a-a03a-11e6-87e6-e940eff65af5.png)

Now here comes the catch. If you look at the exit signpost in real life, **both** ways lead to road 23 (just different sections of it):
![snimek z 2016-11-01 13-56-12](https://cloud.githubusercontent.com/assets/755451/19890450/0cb5a002-a03b-11e6-9158-d96eba5617be.png)
See full view in [google street view](https://www.google.cz/maps/@49.1631642,16.5587258,3a,75y,301.76h,80.96t/data=!3m6!1e1!3m4!1seduSSIePLdj4IKC9pVuu7g!2e0!7i13312!8i6656).

So, in this example, the only good information is actually the `turn:lanes` direction displayed. But turn lanes are tricky, and for more complicated junctions (esp. with several links shortly after each other) they do not serve well.

That basically leaves only two ways of reliable navigation. The first is the exit number. If osmand displayed `exit 190` (which is tagged properly at the [junction point](https://www.openstreetmap.org/node/25566987#map=19/49.16664/16.54819), just not displayed; in the photo above you can see it at the very top of the signpost), it would be a reliable piece of information and it would be easy to see where to turn. However, exit numbers are used only for highways, but not for trunk ways (at least here in Czech Republic).

The second piece of reliable information is the `destination` contents. I don't know why it shows up for you and not for me, but if we can fix it, I'll be very happy. However, I'd prefer to see it in full (the widget going multiline), not just two words and then cut off (as in your screenshot), because sometimes some of those destination names can be shared between several ways and again it could be possibly unreliable in that case (there's not much time to think when in high speed). I understand the full signpost contents can occupy considerable screen space - maybe it could be turned on and off as an option, for people who prefer seeing full direction information? Anything, just seeing in the shorter form would still help a lot.

Thanks in advance!

**Comment by sonora** [comment]
I am not convinced there is any value in "seeing the original sign", because
- many signs are combined and with combined arrows (as your pictures prove),
- and in OSM we have no information what the "original sign" really looks like, so we can only guess from parsing the tags. And this is exactly what we do today:.

Instead, we parse all relevant information from the destination:* tags and display + announce exactly the relevant combination of ref and destination. How much of that fits on your screen is a matter of how big your device screen is, of course, for very small deices you may want to try landscape mode to see the complete text.

Regarding the destination display, see issue #3181 for a status of where we are today. There were some issues, mostly affecting users who selected a fixed map display language and not "Local names" for their maps. All of it should be fixed in nightlies (and next release) except our support for destination:lanes, which is still open as of today.

**Comment by kparal** [comment]
I have set "local names" for map language and enabled TTS for street names, but I still don't see the destination in the top bar. Is there yet another bug? Should I wait until the next release and report a new bug if I still don't see it?

> And this is exactly what we do today

If we can squash the bugs and get it rendered on my display, yes :-) However, it's also about presentation. Clearly seeing the exit signs is one of the reasons why my wife still uses Mapfactor Navigator. It is currently more obvious in there. But that's just a cosmetic issue and a matter of taste, if I can see it rendered at all, anywhere, I personally will be happy.

> How much of that fits on your screen is a matter of how big your device screen is, of course, for very small deices you may want to try landscape mode to see the complete text.

I'm not likely to rotate my phone each time I'm nearing a high way exit (and using the portrait mode feels much more convenient to me for regular driving). That's why I asked for multiline description in the top bar (at least as an option), because losing a few height pixels once in a while (when you're taking an exit) is not a problem when driving on high-speed highways (there's nowhere else to turn anyway).

What about showing the `exit 190` information (in the top bar, or somewhere else)? Is that also rejected? What about replacing the big yellow arrow in top left with a highway exit icon and the exit number? That would be very obvious, very helpful, and would not take any more space on the screen.

**Comment by sonora** [comment]
Exit announcement is  separate issue, not yet implemented.

Regarding the destination display: From my practical experience, the destination:ref (we display that up front) together with the beginning of the destination:(forward/backward) text is the most significant portion of the entire dest information, and that is what is visible even on tiny screens.

The fixes are in our nightly, and will be in the 2.5 release. Cannot say off hand why you do not see any dest info with your current setup. Are you suing the latest maps? Have you looked at the route information screen (cf. screenshot in 3200) and checked if you see dest info in the turn-by-turn instructions?

**Comment by kparal** [comment]
I have the latest maps, I even tried enabling navigation for Live updates. In the route itinerary, I don't see motorway_link's `destination` either. I see there motorway_link's `destination:ref` and motorway_junction's `ref` and `name` (actually both `name:de` and `name` in this order, which is slightly incorrect, because my language is Czech, not German). So no, `destination` is not mentioned in there.

I'll wait for the next release and if it is not fixed, we can debug deeper. Thanks for your time.

**Comment by sonora** [comment]
Yes, I think thd bug you are seeing is fixed in 2.5, which should be out in a few days. I will watch this issue, in case it is not fixed please state here.

**Comment by raindropsfromsky** [comment]
IMHO this feature has tremendous value BUT only in 3D (actually 2.5D) view, where OsmAnd simulates the real world.

So that brings us to the main question: OsmAnd already has 2.5D view in iPhone, but not in Android. Why is this so? 

Many basic Android Navigational apps have the 2.5D view, but the almighty OsmAnd is a remarkable exception!
